apiServer.js

```diff
@@ -19,9 +19,29 @@ const { watch } = require('fs/promises');
 
 const cmdAsync = util.promisify(nodecmd.get);
 const apiPort = userconfig.initial.apiport || config.server.apiport;
-const apiPortHttps = apiPort + 1;
+const apiPortHttps = +apiPort + 1;
 let initialHash = hash(fs.readFileSync(path.join(__dirname, '/config/userconfig.js')));
 
+async function loadUpnpIfRequired() {
+  // User configured UPnP node
+  if (userconfig.initial.apiport || userconfig.initial.routerIP) {
+    if (!userconfig.initial.apiport || !userconfig.initial.routerIP) {
```

Comment:
I think the routerIP actually does not have to be set.

Reply:
Thank you very much for your work/improvements, I am unable to check/test the changes until next week or so but they will not be forgotten.
One of the things that we need to be careful is that having api port doesn't mean it's a UPnP node, as it can be set as 16127 and not be a UPnP node, and from your comments I believe you might have used that as assumption. Correct?
Cheers

Reply:
> You're right - the routerIP is only needed for the firewall adjustment if UPnP is configured, have adjusted.
> 
> In a future PR, I would like to rework some of the natupnp module. This is the biggest area of problems users have IMO.
> 
> 1/ Add better logging around errors. 2/ Add checks to ensure multicast is working (Join 239 group and issue "self ping") for warning messages, there will be no response if snooping is enabled and filtering. 3/ Listen for igmp queries 4/ Cache SSDP response in case of IMGP snooping. This will fix most problems, as users who have misconfigured networks (igmp snooping with no querier) the bridge / switch will allow SSDP to work for the first 5 minutes (depending on switch) before the igmp join times out. After that, for the actual upnp part - multicast isn't required. 5/ Get router IP from SSDP response for firewall - then user doesn't have to specify this in the config

I checked this again - routerIP needs to be set or the firewall won't be opened to allow UPnP

Reply:
Yes - had assumed that apiPort set in userConfig.js meant UPnP, my bad.

So from what I can see, currently it works like this:

If apiPort is set, and it's not 16127, then node is UPnP
If apiPort is set and it's 16127 AND routerIP is set it's UPnP.
if apiPort is set and it's 16127 and no routerIP - then it's a normal node.

If you can confirm the above is correct - I'll fix the code to follow the above.

Might be a good idea to add a specific isupnpnode: true/false to the config for the future? but I could see that taking years to migrate to.

I've got some ideas around auto config of upnp nodes using multicast - nodes can join a group on the lan to discover what nodes already exist, then they can allocate there own api ports amongst themselves.

Cheers!

Reply:
If router IP it's UPnP, if apiport different from 16127 it's UPnP, if apiport equal to 16127 we check if it's UPnP but if it's not fluxos doesn't shutdown.